Introduce functional tests for new REST test plugins

jakelandis opened the pull request:
This commit introduces 2 new Gradle plugin functional tests. 
These tests make use of dummy subprojects to help satisfy the
inter dependencies of the projects in use. Helper methods
have been added to the shared common parent.

jakelandis committed:
introduce functional test for yamlRestTest

jakelandis committed:
cleanup

jakelandis committed:
add test for yamlRestCompatTest

jakelandis commented on buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java:
```diff
@@ -73,9 +73,11 @@ public void apply(Project project) {
         GradleUtils.extendSourceSet(project, YamlRestTestPlugin.SOURCE_SET_NAME, SOURCE_SET_NAME);
 
-        // create the test cluster container, and always use the default distribution
+        // create the test cluster container, and use the default distribution. System property is only used to help test this plugin.
```
Wasn't a fan of doing this...but for reasons I have hard time following, this requires the `:distribution:archives:linux-tar` (my workstation is a linux box) project to exist. I assume that OS specific, and did not want to have to mirror all of the variants from distribution:archive as dummy projects. 

Alternatively, I think i can only run this test specifically for a given OS, but avoided that since there is some file copying going on and wanted to get coverage across OS's. The distribution in use is not an important detail of the test.

breskeby replied:
I mentioned in another comment that you should be able to just configure the distribution explicitly in the tests so this sys property in the production code should not be required. IMO its fine to not test all OS specific resolutions as this is mostly logic done in the distribution-download-plugin and the test cluster plugin which we test separately

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

breskeby commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/YamlRestCompatTestPluginFuncTest.groovy:
```diff
@@ -0,0 +1,111 @@
+        setupInternalRestResources(api, test)
+
+        //don't use default distro from here since it requires resolving OS specific project and don't want to create the mock projects
+        System.setProperty("tests.rest.compat.use.default.distro", "false");
```
you should be able to get rid of the need for this property by just explicitly configure the test distribution here in the test like adding:
sth like
```
        buildFile << """
        testClusters {
            yamlRestCompatTest.setTestDistribution(TestDistribution.INTEG_TEST)
        }
        """
```

breskeby replied:
this needs to be added after the plugin is applied though otherwise the test cluster doesn't exist yet

jakelandis replied:
thanks! fixed in https://github.com/elastic/elasticsearch/pull/64186/commits/249f64f05d572de4c0480ab87fae858ee7902442 , much better then the System property.

breskeby reviewed (changes requested):
One change regarding configuring tests to use the test distribution. Otherwise this looks good to me. glad we start adding more integration tests here

jakelandis committed:
review comments

jakelandis commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/YamlRestTestPluginFuncTest.groovy:
```diff
@@ -0,0 +1,92 @@
+        """
+
+        when:
+        def result = gradleRunner("yamlRestTest", '-i').build()
```
removed it from tests that don't read from the output.

rjernst replied:
This was necessary in the past so that we get anything meaningful from failures. Without it, the inner build had essentially no output.

jakelandis commented:
@breskeby - Thanks for the review, comments addressed and should be ready for another round.

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into yaml_rest_test_tests

mark-vieira commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/YamlRestCompatTestPluginFuncTest.groovy:
```diff
@@ -0,0 +1,114 @@
+
+        File resourceDir = new File(testProjectDir.root, "build/resources/yamlRestCompatTest/rest-api-spec")
+        // ensure the compatible resources from bwc:minor are in the resource dir
+        assert new File(resourceDir, "test/foo/10_basic.yml").exists()
```
I don't believe we need `assert` here. Any boolean expression in a `then:` block is an implicit assertion.

mark-vieira commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/YamlRestCompatTestPluginFuncTest.groovy:
```diff
@@ -0,0 +1,114 @@
+
+        result = gradleRunner("yamlRestCompatTest", '-i').buildAndFail() //expect failure - we don't actually spin up a cluster
+
+        then:
```
What exactly are we testing in this scenario? Just that the copy tasks are still up to date? If not, should this be a separate feature method?

jakelandis replied:
The first `when` is testing that copy is correct, and the second `when` that test attempts to spin up a cluster and execute. 

Testing that the copy is up to date is probably more about the behavior of the copy resource plugin, not this plugin. (We don't have tests for that yet, but we probably should, just not in this PR).

If you think is a confusing, i can move the config from the second `when` up to the first, and not test the up to date behaviour here.

mark-vieira replied:
Yeah, I'm wondering if this makes for a good integration test given that it's actually failing to do so. We're just testing that it "attempts" to do so and that might be quicker and easier to validate in a unit test. That is use something like `ProjectBuilder` to validate that applying the plugin does what we expect, like creates the source set, the corresponding test cluster, etc.

@breskeby thoughts?

breskeby replied:
yeah that might be better tested in a unit test if we think its required

mark-vieira commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/YamlRestTestPluginFuncTest.groovy:
```diff
@@ -0,0 +1,92 @@
+        }
+        """
+
+        result = gradleRunner("yamlRestTest", '-i').buildAndFail() //expect to fail since we don't actually spin up a cluster
```
This seems the same as the last one. What exactly is this testing? Are we just testing that the source set is created properly?

mark-vieira reviewed (changes requested):
Thanks for additing this additional coverage, Jake! Just a few questions.

breskeby commented:
thanks for addressing my feedback. I have nothing more to add to what @mark-vieira added.

breskeby reviewed (approved):
lgtm after marks comments are addressed

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into yaml_rest_test_tests

jakelandis committed:
Merge branch 'master' of github.com:elastic/elasticsearch into yaml_rest_test_tests

jakelandis committed:
wip: add rest resource plugin test (broke others)

jakelandis committed:
fix other tests

jakelandis committed:
minor review change

jakelandis commented:
@mark-vieira @breskeby - ready for another round of reviews (hopefully the last)

I reworked the tests here and introduced tests specific to the rest resources plugin as well.  The test no longer expects to build and fail, and covers more scenarios more inline with functional testing.

mark-vieira reviewed (approved):
I did a very cursory look and this looks good. Thanks @jakelandis.

@breskeby do you mind taking another look at this as well?

breskeby commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/RestResourcesPluginFuncTest.groovy:
```diff
@@ -0,0 +1,185 @@
+        result.task(':copyRestApiSpecsTask').outcome == TaskOutcome.SUCCESS
+        result.task(':copyYamlTestsTask').outcome == TaskOutcome.SUCCESS
+        File resourceDir = new File(testProjectDir.root, "build/resources/test/rest-api-spec")
+        new File(resourceDir, "/api/" + apiCore1).exists()
```
just a little hint. 

you could use the convenience `file()` method for better reading:
```
file("/build/resources/test/rest-api-spec/api/" + apiCore1).exists()
```
even though there's some repeating I personally think this makes the test more readable.

jakelandis replied:
done https://github.com/elastic/elasticsearch/pull/64186/commits/e02be1a384dbb1e25596fbfd5c5f48b0af36d1d0

breskeby commented on buildSrc/src/integTest/groovy/org/elasticsearch/gradle/fixtures/AbstractGradleFuncTest.groovy:
```diff
@@ -136,4 +142,47 @@ abstract class AbstractGradleFuncTest extends Specification {
         }
     }
+
+    void setupRestResources(List<String> apis, List<String> tests = [], List<String> xpackApis = [], List<String> xpackTests = []) {
```
I think this would better live in an AbstractRestResourcesFuncTest than in the general AbstractGradleFuncTest. `AbstractGradleFuncTest` should only contain more general logic applied to a most functional tests

jakelandis replied:
done: https://github.com/elastic/elasticsearch/pull/64186/commits/5021379eabe4a2362a763df91d342e8bbc3af495

breskeby reviewed (approved):
just two minor remarks

jakelandis committed:
introduce AbstractGradleFuncTest

jakelandis committed:
prefer file(..) over new File()
